imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -321,23 +202,32 @@ def science_direct_event(packets_data: xr.Dataset) -> xr.Dataset:
     dataset : xarray.Dataset
         Xarray dataset.
     """
-    de_data_list = []
-    packet_met_time = []
+    de_data_dict: dict[str, list] = defaultdict(list)
 
     # Because DE_TOF is a variable length data,
     # I am using extend to add another list to the
     # end of the list. This way, I don't need to flatten
     # the list later.
     for i, data in enumerate(packets_data["de_tof"].data):
-        binary_str_val = convert_to_binary_string(data)
-        # break binary stream data into unit of 48-bits
-        event_48bits_list = break_into_bits_size(binary_str_val)
-        # parse 48-bits into meaningful data such as metaevent or direct event
-        de_data_list.extend([parse_direct_event(event) for event in event_48bits_list])
-        # add packet time to packet_met_time
-        packet_met_time.extend(
-            [packets_data["shcoarse"].data[i]] * len(event_48bits_list)
-        )
+        parsed_de_data = parse_direct_events(data)
+        for key, new_data in parsed_de_data.items():
+            de_data_dict[key].extend(new_data)
+
+        # add packet data to keep_packet_data dictionary, repeating values
+        # for each direct event encoded in the current packet
+        for from_key, to_key in {
+            "shcoarse": "ccsds_met",
+            "src_seq_ctr": "src_seq_ctr",
+            "pkt_len": "pkt_len",
+            "last_spin_num": "last_spin_num",
+            "spin_invalids": "spin_invalids",
+            "esa_step_num": "esa_step",
+            "meta_seconds": "meta_seconds",
+            "meta_subseconds": "meta_subseconds",
+        }.items():
+            de_data_dict[to_key].extend(
+                [packets_data[from_key].data[i]] * len(parsed_de_data["de_tag"])
```

Comment:
What does this line do? Just curious

Reply:
Haha, yeah. I can comment that. It is actually a remnant from what you wrote originally. It just makes a list of identical values that is the length of the number of parsed DEs.